### ISSUE TITLE
Tests failing against latest jruby

### ISSUE THREAD

**Issue by deivid-rodriguez** [brigade/overcommit#343]
I tried to run the test suite against latest JRuby (9.0.5.0) and got some failing tests. It seemed like an order dependent issue since the tests pass when run in isolation, so I bisected the failures. 

```
$ jruby --version
jruby 9.0.5.0 (2.2.3) 2016-01-26 7bee00d OpenJDK 64-Bit Server VM 24.95-b01 on 1.7.0_95-b00 [linux-amd64]
```
```
$ bundle exec rspec ./spec/overcommit/hook/pre_push/protected_branches_spec.rb[1:1:1:1,1:1:2:1,1:2:1:1:1,1:2:1:2:1,1:2:2:1:1,1:2:2:2:1]
Coverage may be inaccurate; Try setting JRUBY_OPTS="-Xcli.debug=true --debug"
Run options: include {:ids=>{"./spec/overcommit/hook/pre_push/protected_branches_spec.rb"=>["1:1:1:1", "1:1:2:1", "1:2:1:1:1", "1:2:1:2:1", "1:2:2:1:1", "1:2:2:2:1"]}}
.FFFFF

Failures:

  1) Overcommit::Hook::PrePush::ProtectedBranches when pushing to unprotected branch when push is destructive should pass the check
     Failure/Error: protected?(pushed_ref.remote_ref) && pushed_ref.destructive?

     NameError:
       undefined method `remote_ref' for class `#<Class:0x4772b4a6>'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:18:in `block in illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:17:in `illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:5:in `run'
     # ./spec/support/matchers/hook.rb:10:in `matches?'
     # ./spec/support/matchers/hook.rb:84:in `block in matches?'
     # ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:36:in `block in (root)'
     # ./spec/spec_helper.rb:52:in `block in (root)'
     # ./lib/overcommit/utils.rb:259:in `with_environment'
     # ./spec/spec_helper.rb:51:in `block in (root)'

  2) Overcommit::Hook::PrePush::ProtectedBranches when pushing to protected branch when branch name matches a protected branch exactly when push is not destructive should pass the check
     Failure/Error: protected?(pushed_ref.remote_ref) && pushed_ref.destructive?

     NameError:
       undefined method `remote_ref' for class `#<Class:0x95e004c>'
     Shared Example Group: "protected branch" called from ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:61
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:18:in `block in illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:17:in `illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:5:in `run'
     # ./spec/support/matchers/hook.rb:10:in `matches?'
     # ./spec/support/matchers/hook.rb:84:in `block in matches?'
     # ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:46:in `block in (root)'
     # ./spec/spec_helper.rb:52:in `block in (root)'
     # ./lib/overcommit/utils.rb:259:in `with_environment'
     # ./spec/spec_helper.rb:51:in `block in (root)'

  3) Overcommit::Hook::PrePush::ProtectedBranches when pushing to protected branch when branch name matches a protected branch exactly when push is destructive should fail
     Failure/Error: protected?(pushed_ref.remote_ref) && pushed_ref.destructive?

     NameError:
       undefined method `remote_ref' for class `#<Class:0x377c0e2a>'
     Shared Example Group: "protected branch" called from ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:61
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:18:in `block in illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:17:in `illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:5:in `run'
     # ./spec/support/matchers/hook.rb:10:in `matches?'
     # ./spec/support/matchers/hook.rb:63:in `block in matches?'
     # ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:54:in `block in (root)'
     # ./spec/spec_helper.rb:52:in `block in (root)'
     # ./lib/overcommit/utils.rb:259:in `with_environment'
     # ./spec/spec_helper.rb:51:in `block in (root)'

  4) Overcommit::Hook::PrePush::ProtectedBranches when pushing to protected branch when branch name matches a protected branch glob pattern when push is not destructive should pass the check
     Failure/Error: protected?(pushed_ref.remote_ref) && pushed_ref.destructive?

     NameError:
       undefined method `remote_ref' for class `#<Class:0x7fa02e81>'
     Shared Example Group: "protected branch" called from ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:66
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:18:in `block in illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:17:in `illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:5:in `run'
     # ./spec/support/matchers/hook.rb:10:in `matches?'
     # ./spec/support/matchers/hook.rb:84:in `block in matches?'
     # ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:46:in `block in (root)'
     # ./spec/spec_helper.rb:52:in `block in (root)'
     # ./lib/overcommit/utils.rb:259:in `with_environment'
     # ./spec/spec_helper.rb:51:in `block in (root)'

  5) Overcommit::Hook::PrePush::ProtectedBranches when pushing to protected branch when branch name matches a protected branch glob pattern when push is destructive should fail
     Failure/Error: protected?(pushed_ref.remote_ref) && pushed_ref.destructive?

     NameError:
       undefined method `remote_ref' for class `#<Class:0x5a06f4ba>'
     Shared Example Group: "protected branch" called from ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:66
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:18:in `block in illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:17:in `illegal_pushes'
     # ./lib/overcommit/hook/pre_push/protected_branches.rb:5:in `run'
     # ./spec/support/matchers/hook.rb:10:in `matches?'
     # ./spec/support/matchers/hook.rb:63:in `block in matches?'
     # ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:54:in `block in (root)'
     # ./spec/spec_helper.rb:52:in `block in (root)'
     # ./lib/overcommit/utils.rb:259:in `with_environment'
     # ./spec/spec_helper.rb:51:in `block in (root)'

Finished in 0.19 seconds (files took 0.539 seconds to load)
6 examples, 5 failures

Failed examples:

rspec ./spec/overcommit/hook/pre_push/protected_branches_spec.rb:36 # Overcommit::Hook::PrePush::ProtectedBranches when pushing to unprotected branch when push is destructive should pass the check
rspec ./spec/overcommit/hook/pre_push/protected_branches_spec.rb[1:2:1:1:1] # Overcommit::Hook::PrePush::ProtectedBranches when pushing to protected branch when branch name matches a protected branch exactly when push is not destructive should pass the check
rspec ./spec/overcommit/hook/pre_push/protected_branches_spec.rb[1:2:1:2:1] # Overcommit::Hook::PrePush::ProtectedBranches when pushing to protected branch when branch name matches a protected branch exactly when push is destructive should fail
rspec ./spec/overcommit/hook/pre_push/protected_branches_spec.rb[1:2:2:1:1] # Overcommit::Hook::PrePush::ProtectedBranches when pushing to protected branch when branch name matches a protected branch glob pattern when push is not destructive should pass the check
rspec ./spec/overcommit/hook/pre_push/protected_branches_spec.rb[1:2:2:2:1] # Overcommit::Hook::PrePush::ProtectedBranches when pushing to protected branch when branch name matches a protected branch glob pattern when push is destructive should fail

Coverage report generated for RSpec to /home/deivid/Code/overcommit/coverage. 978 / 2290 LOC (42.71%) covered.
```

**Comment by sds** [comment]
Thanks for the report, @deivid-rodriguez. This error is a bit odd, as it's complaining about a missing field on a struct (`pushed_refs` returns an array of `PushedRef` instances).

I don't work with any version of JRuby regularly, so I'm not sure why it's representing a Struct as a Class in this case (or perhaps JRuby is just displaying it as `#<Class:0x4772b4a6>` while MRI outputs `<#struct PushedRef ...>`).

In fact, the only reason Overcommit even supports JRuby was because of a request a long time ago from an engineer at @causes who wanted to be able to use it in an experimental project (which never went anywhere). My experience with JRuby is that it is terrible for running hooks because of the slow startup time.

Anyway, all that is to say I'm not sure when I'll get time to have a look at this, given JRuby is not a runtime I ever use.

**Comment by deivid-rodriguez** [comment]
@sds I'd say this is a JRuby issue, so I've opened an issue there: https://github.com/jruby/jruby/issues/3686

**Comment by sds** [comment]
Cool, thanks for reporting the issue upstream.

They seem to have triaged it and marked it as getting fixed for JRuby 9.1.0.0, so I'm going to close.

**Comment by deivid-rodriguez** [comment]
@sds Just for your information, this was an issue with rspec-mocks and should be fixed in the next release!
